### PR TITLE
[TTAHUB-2369] Admin course import FE

### PR DESCRIPTION
**Pull request by AdamAdHocTeam**
## Description of change

Add a course import to to the admin menu.

## How to test

- Review all code.
- All tests pass.
- Test the three cases for course import:
1. **Creates**: Course doesn't exist in db.
2. **Updates**: Course already exists in db with and without (spacing special chars are diff) exact wording , updated at is changed.
3. **Marks deleted**: Course pre existed in the db but is no longer in the import.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2369

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/pages/Admin/components/CsvImport.js [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+      importIds.shift();
+
+      // Remove any blank values.
+      importIds = importIds.filter((id) => id !== '');
+
+      const distinctImportIds = [...new Set(importIds)];
+
+      // Check if we have duplicate primary Id Column values.
+      if (distinctImportIds.length !== importIds.length) {
+        setError(`Duplicate ${primaryIdColumn}'s found. Please correct and try again.`);
```
Comment (thewatermethod) [inline frontend/src/pages/Admin/components/CsvImport.js:133]:
```suggestion
        setError(`Duplicate ${primaryIdColumn}s found. Please correct and try again.`);
```

**Comment by thewatermethod** [comment]
Really like the generic component approach, I think that makes a lot of sense. Worked well when I was testing it.

Couple of thoughts:

As is, this seems like it would be pretty annoying to use. The file attached to the Jira, for example, failed because it contained a duplicate, but it gave no feedback on what the duplicate was. Do we want to fail on duplicate or just skip? If we fail, do we want to identify what the duplicate was? I know that we are trying to set our customer up to maintain this, but we should keep in mind that we'll be supporting it if it becomes a struggle.

Also, are we sure that we want to delete all the courses that don't appear on the CSV when it is uploaded? I assume you checked, I don't see that in the Jira specs.

File: src/migrations/20231206210453-update-flags-for-user.js [hunk 2]
```diff
@@ -6,7 +6,7 @@ module.exports = {
     await queryInterface.sequelize.transaction(async (transaction) => {
       const sessionSig = __filename;
       await prepMigration(queryInterface, transaction, sessionSig);
-      return updateUsersFlagsEnum(queryInterface, transaction, ['training_reports']);
+      // return updateUsersFlagsEnum(queryInterface, transaction, ['training_reports']);
```
Comment (thewatermethod) [inline src/migrations/20231206210453-update-flags-for-user.js:9]: this shouldn't be in this PR, should it?

File: src/constants.js [hunk 3]
```diff
@@ -223,6 +223,7 @@ const FEATURE_FLAGS = [
   'goal_source',
   'merge_goals',
   'monitoring',
+  'training_reports',
```
Comment (thewatermethod) [inline src/constants.js:226]: this shouldn't be in this PR, should it?

File: src/services/course.test.js [hunk 4]
```diff
@@ -0,0 +1,275 @@
+import { Op } from 'sequelize';
+import { expect } from '@playwright/test';
```
Comment (thewatermethod) [inline src/services/course.test.js:2]: I don't think we want to import this here

File: frontend/src/pages/Admin/components/CsvImport.js (hunk 1 above)
Comment (AdamAdHocTeam) [reply frontend/src/pages/Admin/components/CsvImport.js]: Fixed.

**Comment by AdamAdHocTeam** [comment]
> Really like the generic component approach, I think that makes a lot of sense. Worked well when I was testing it.
> 
> Couple of thoughts:
> 
> As is, this seems like it would be pretty annoying to use. The file attached to the Jira, for example, failed because it contained a duplicate, but it gave no feedback on what the duplicate was. Do we want to fail on duplicate or just skip? If we fail, do we want to identify what the duplicate was? I know that we are trying to set our customer up to maintain this, but we should keep in mind that we'll be supporting it if it becomes a struggle.
> 
> Also, are we sure that we want to delete all the courses that don't appear on the CSV when it is uploaded? I assume you checked, I don't see that in the Jira specs.

As we discussed message for duplicates has been updated. Also the BE ticket outlines marking deleted when course no longer appears in the list.

File: src/services/course.test.js (hunk 4 above)
Comment (AdamAdHocTeam) [reply src/services/course.test.js]: Good call.

File: src/migrations/20231206210453-update-flags-for-user.js (hunk 2 above)
Comment (AdamAdHocTeam) [reply src/migrations/20231206210453-update-flags-for-user.js]: There were actually two commented out migrations. Not sure why but was having issues updating db.

File: src/constants.js (hunk 3 above)
Comment (AdamAdHocTeam) [reply src/constants.js]: I think this was because my db already expected it. Will remove.

**Review by nvms — COMMENTED** [review]
Minor comments

File: src/services/course.js [hunk 5]
```diff
@@ -19,9 +19,12 @@ export async function csvImport(buffer) {
 
   const parsed = parse(buffer, { skipEmptyLines: true, columns: true });
   let rowCount = 1;
-  const results = parsed.map(async (course) => {
-    try {
-      let rawCourseName = course['course name'];
+  let results;
+  try {
+    results = await Promise.all(parsed.map(async (course) => {
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/services/course.js:24]: Will look to make this in the BE PR: https://github.com/HHS/Head-Start-TTADP/pull/1878

File: src/services/course.js [hunk 6]
```diff
@@ -19,9 +19,12 @@ export async function csvImport(buffer) {
...
+      // Get the first property value form course object.
+      let rawCourseName = Object.values(course)[0];
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/services/course.js:26]: done in other pr.

File: src/services/course.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/services/course.js:24]: No need to make this change end point is wrapped in txn. Also bad practice per sql doc according to @thewatermethod.

**Comment by thewatermethod** [comment]
looks good, should the target branch for this be the backend PR instead of main?

**Comment by AdamAdHocTeam** [comment]
> looks good, should the target branch for this be the backend PR instead of main?

Done.

**Review by thewatermethod — APPROVED** [review]

**Pull request merged**